[MRG+1] Fix Bayesian ridge regression

albertcthomas opened the pull request:
#### Reference Issues/PRs
This PR is taking over PR #11294 (duplicate of PR #10751) which is fixing the sign of log(det(Sigma)) in the computation of the marginal log likelihood in BayesianRidge.
Fixes issue #10748


#### What does this implement/fix? Explain your changes.
In order to implement a non regression test for the score I took some time to understand the code and I found some issues about the returned attributes and the doc.

- The references given in the doc and the docstring are 'Bayesian interpolation' (MacKay, 1992) and slides from a lecture which seems to follow 'Pattern Recognition and Machine learning' by Bishop. However the scikit-learn implementation of Bayesian ridge regression (using Gamma distribution for the hyperpriors) is just Automatic Relevance determination with the same hyperparameter `lambda` for all coordinates of `w`. Hence IMO the best reference to understand the code (and the one that I found to be the most helpful) is the one given for Automatic Relevance determination, i.e. , 'Sparse Bayesian Learning and the Relevance Vector Machine' by Tipping. So I would add this reference to the docstring and the doc of Bayesian ridge regression.

- The docstring of the `scores_` attribute reads : `value of the objective function (to be maximized)`. I think we should be clearer about what the objective function is here, i.e., the marginal log likelihood and maybe give the formula in the doc (equation (36) in the paper by Tipping) to be clear about what is done in the scikit learn implementation

![objective_function](https://user-images.githubusercontent.com/15966638/46095062-ea5feb80-c1bb-11e8-98c0-f5917694515c.PNG)

- The returned `coef_` attribute does not correspond to the last update of `alpha_` and `lambda_` whereas the `sigma_` matrix is the one corresponding to the last update. I think this is a bug for `coef_` as when doing prediction the `coef_` attribute should be the one updated thanks to the last update values of `alpha_` and `lambda_` (see equation (19), (21) and (22) in Tipping):
![prediction](https://user-images.githubusercontent.com/15966638/46095173-27c47900-c1bc-11e8-9c7c-5bb698dd645e.PNG)

Besides it was decided to return the `alpha_` and `lambda_` parameters corresponding to the (non updated) `coef_` attribute and not the last updates of `alpha_` and `lambda_` in issue #8224. I think this is wrong as we should also return the last updates because there are the ones we should use in the posterior. If it is agreed that we should update `coef_` as explained above, the last updates of `alpha_` and `lambda_` will correspond to the updated `coef_` value.

- The first test in `test_bayes` is skipped because it is broken and I think we should fix it or remove it but I am not certain about the correct output yet, maybe @agramfort can help for this one? It checks that the scores are increasing.


- [x] Fix sign of logdet Sigma in the score (previous PRs)
- [x] improve reading of the code: variable `sigma` is now named `scaled_sigma` and update some of the comments
- [x] Non regression tests
- [x] Update doc and refs
- [x] Remove skipped increasing scores test.
- [x] Fix `coef_`, `lambda_` and `alpha_` attributes.


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->
Closes  #11294 #10751 #10748

bartz committed:
Fixes sign of logdet_sigma term in BayesianRidge.fit

albertcthomas committed:
improve reading of the code

albertcthomas committed:
sc in score

albertcthomas committed:
comment about gamma value

agramfort commented:
woooottt I have not looked at this code for the last 8 years :) (git blame will confirm...)

if the skipped test now passes with your changes then you have a non-regression test...

thx @albertcthomas for looking into this

albertcthomas commented:
The scikit-learn implementation uses the MacKay updates and according to section 2.2 in [A New View of Automatic Relevance Determination (Wipf and Nagarajan, 2008)](http://papers.nips.cc/paper/3372-a-new-view-of-automatic-relevance-determination.pdf) "MacKay updates do not even guarantee cost function decrease" so the skipped test should be removed.

agramfort commented:
ok can you just insert comments in the code to say this? thx

albertcthomas committed:
remove skipped test because not well founded + add explanations and refs in docstring

albertcthomas committed:
fix alpha, lambda and coef output values

albertcthomas commented:
Ok I removed the skipped test, added details to the docstring about the implementation and how the parameters are updated, and fixed the output values for `coef_`, `alpha_` and `lambda_`. It should now be possible to add a non regression test for the score computation (main motivation of this PR :)). Will do that soon, as well as adding a few more details in the doc.

albertcthomas committed:
change score position

albertcthomas committed:
refactor computation of score

albertcthomas committed:
add test for score and add a check for n_iter value

albertcthomas committed:
clarify doc and add ref

albertcthomas committed:
delete newlines

albertcthomas committed:
fix doc

agramfort commented on sklearn/linear_model/bayes.py:
```diff
@@ -110,18 +119,21 @@ class BayesianRidge(LinearModel, RegressorMixin):
     -----
-    For an example, see :ref:`examples/linear_model/plot_bayesian_ridge.py
-    <sphx_glr_auto_examples_linear_model_plot_bayesian_ridge.py>`.
+    There exist several strategies to perform Bayesian ridge regression. This
```
exist -> exists

albertcthomas replied:
I think 'exist' is correct here.

agramfort reviewed (approved):
besides my nitpick LGTM

thx @albertcthomas

albertcthomas commented:
Thanks for the review @agramfort

TomDLT commented on sklearn/linear_model/bayes.py:
```diff
@@ -217,45 +236,62 @@ def fit(self, X, y, sample_weight=None):
+        # return regularization parameters
+        self.alpha_ = alpha_
+        self.lambda_ = lambda_
+        # and corresponding posterior mean and posterior covariance
```
Here you recompute the posterior mean and posterior covariance, exactly as in the for-loop.
Could you refactor the two computations into a function, to avoid duplication?
And don't you need to update `logdet_sigma_` also?

albertcthomas replied:
> And don't you need to update `logdet_sigma_` also?

Indeed and and `rmse_` as well. Thanks!

TomDLT commented:
You'll need a whats new entry.

albertcthomas committed:
refactor code and fix last score

albertcthomas commented:
Thanks for the review @TomDLT.

albertcthomas committed:
whatsnew entry

TomDLT committed:
Change PR number in whats new

albertcthomas commented:
@TomDLT you put the PR number instead of the issue number in what’s new.

jnothman commented:
> you put the PR number instead of the issue number in whatsnew.

We usually do.

albertcthomas commented:
Ah ok it makes sense actually :). thanks @TomDLT

jnothman committed:
Merge branch 'master' into fix_sign_bayes_ridge

jnothman commented:
Thanks for the ping @auvipy!

auvipy commented:
wc!!
